File: src/path-matcher.test.ts
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { PathMatcher } from "./path-matcher";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

interface TestCase {
	description: string;
	pattern: string;
	path: string;
	match: boolean;
}

describe("Path Matcher", () => {
	const fixturesPath = path.join(
		__dirname,
		"../fixtures/path_matcher_fixtures.json",
	);
	const fixtures: TestCase[] = JSON.parse(
		fs.readFileSync(fixturesPath, "utf8"),
	);

	const pathMatcher = new PathMatcher();

	fixtures.forEach((testCase) => {
		const {
			description,
			pattern,
			path: filePath,
			match: shouldMatch,
		} = testCase;
		const testName = `${description}: '${pattern}' should ${shouldMatch ? "match" : "NOT match"} '${filePath}'`;

		it(testName, () => {
			const result = pathMatcher.match(filePath, pattern);

			expect(result).toBe(shouldMatch);
		});
	});
});

File: test/path-matcher.test.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'

fixtures_path = File.join(__dir__, 'path_matcher_fixtures.json')
fixtures = JSON.parse(File.read(fixtures_path))

puts "--- Testing Pattern Matching Rules ---"

failed = false

fixtures.each do |test_case|
  pattern = test_case['pattern']
  path = test_case['path']
  should_match = test_case['match']
  description = test_case['description'] || pattern

  matches = File.fnmatch?(pattern, path, File::FNM_PATHNAME | File::FNM_DOTMATCH)

  if matches == should_match
    # success
    puts "PASS: #{description} - '#{pattern}' against '#{path}'"
  else
    puts "FAIL: #{description} - '#{pattern}' against '#{path}'"
    puts "  Expected match: #{should_match}, Got: #{matches}"
    failed = true
  end
end

if failed
  exit 1
else
  puts "All tests passed."
end
